### PR TITLE
Content image srcsets

### PR DESCRIPTION
**Pull request by rich-nguyen**
This updates all the remaining uses of imager to the new srcset/picturefill system. Had to do it in one go really. Roughly:
- Use a breakpoint-based system for content (ie. non-facia) responsive images. A breakpoint may have an image width associated, which specifies the dimension of the source image at that breakpoint.
- Remove Imager, and use picturefill, or picturefill+lazyload.
- Reduce the number of profiles, as most are used for image picking (now done by srcset/sizes).
- Rewrite image cleaner so that it uses less jsoup, more html fragments.
- Update Gallery and Lightbox to use srcset.

13253785265_b865a574ef_b

### CODE REVIEW

**Comment by paperboyo** [comment]
:gem:!

un chien interrompu
[https://www.youtube.com/watch?v=1YhmrD7J7ZI&t=71&hd=1]

File: applications/app/views/fragments/galleryBody.scala.html [hunk 1]
```diff
@@ -42,38 +44,29 @@ <h2 class="fc-container__header__title">
...
+                data-link-name="Launch Gallery Lightbox" data-is-ajax>
 
-                    <a class="gallery2__img-container js-gallerythumbs js-delayed-image-upgrade @if(image.width >= image.height) {gallery2__img-container--landscape} else {gallery2__img-container--portrait}"
-                        href="@LinkTo{@gallery.url#img-@rowNum}"
-                        data-src="@imgSrc"
-                        data-link-name="Launch Gallery Lightbox" data-is-ajax>
+                <img class="gallery2__img js-gallery-img @if(image.width >= image.height) {gallery2__img--landscape} else {gallery2__img--portrait}"
+                    srcset="@ImgSrc.srcset(imageElement, GalleryMedia.Inline)"
+                    sizes="@GalleryMedia.Inline.sizes"
+                    src="@ImgSrc.getFallbackUrl(imageElement)"
```
Comment (robertberry) [inline applications/app/views/fragments/galleryBody.scala.html]: Giving a fallback src will mean any browsers that need to be picturefilled (which is most browsers atm) will download that image, too. If this is above the fold that might be what you want anyway - if it's not, you should remove it.

File: common/app/layout/WidthsByBreakpoint.scala [hunk 2]
```diff
@@ -87,21 +86,127 @@ object WidthsByBreakpoint {
...
+
+    val Lightbox = WidthsByBreakpoint(
+      mobile =          Some(465.px),
+      mobileLandscape = Some(645.px),
+      phablet =         Some(725.px),
+      tablet =          Some(965.px),
+      desktop =         Some(1065.px),
+      leftCol =         Some(1225.px),
+      wide =            Some(1920.px))
+  }
```
Comment (robertberry) [inline common/app/layout/WidthsByBreakpoint.scala]: Wowsa, a lot more variations of widths in Content. You might want to add Profiles for these if you've not ...

File: common/app/model/content.scala [hunk 3]
```diff
@@ -720,7 +719,9 @@ trait Lightboxable extends Content {
           "caption" -> JsString(img.caption.getOrElse("")),
           "credit" -> JsString(img.credit.getOrElse("")),
           "displayCredit" -> JsBoolean(img.displayCredit),
-          "src" -> JsString(ImgSrc(img.url.getOrElse(""), ImgSrc.Imager)),
+          "src" -> JsString(ImgSrc.findSrc(imgContainer, Item700).getOrElse("")),
```
Comment (robertberry) [inline common/app/model/content.scala]: Same as above comment

File: common/app/views/fragments/contentImage.scala.html [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+@import views.support.ImgSrc
+
+<figure itemprop="associatedMedia image" itemscope itemtype="http://schema.org/ImageObject"
+        class="element element-image @figureClasses"
+        @image.mediaId.map{ id => data-media-id="@id" }>
+
+    <img srcset="@ImgSrc.srcset(imageContainer, widthsByBreakpoint)"
+         sizes="@widthsByBreakpoint.sizes"
+         src="@ImgSrc.getFallbackUrl(imageContainer)"
```
Comment (robertberry) [inline common/app/views/fragments/contentImage.scala.html]: Same as above comment

File: common/app/views/fragments/javaScriptLaterSteps.scala.html [hunk 5]
```diff
@@ -283,13 +283,13 @@
...
-                    images.upgradePicture();
+                    images.upgradePictures();
                     if(!document.addEventListener) { // IE8 and below
-                        window.onload = images.upgradePicture;
+                        window.onload = images.upgradePictures;
                     }
                 } else {
                     lazyLoadImages.setUpListeners();
-                    images.upgrade();
+                    images.upgradePictures();
```
Comment (robertberry) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: Given this happens on either branch, might as well just move it outside the if statement.

File: common/app/views/support/Profile.scala [hunk 6]
```diff
@@ -57,100 +57,21 @@ case class VideoProfile(
...
-    Item460,
-    Item620,
-    Item640,
-    Item700,
-    Item940,
-    SeoOptimisedContentImage
-  )
-
-  // image widths available to <picture> and <img srcset>
-  lazy val images: Seq[Profile] = Seq(
```
Comment (robertberry) [inline common/app/views/support/Profile.scala]: ?

File: common/app/layout/WidthsByBreakpoint.scala [hunk 7]
```diff
@@ -132,6 +234,12 @@ case class WidthsByBreakpoint(
 
     case _ => Seq.empty
   }
+
+  val profiles: Seq[Profile] = allWidths
```
Comment (robertberry) [inline common/app/layout/WidthsByBreakpoint.scala]: Oh wow, this is a good idea.

File: common/app/views/support/Profile.scala (hunk 6 above)
Comment (robertberry) [inline common/app/views/support/Profile.scala]: Ignore this, I've seen what you've done above now.

**Comment by robertberry** [comment]
Brilliant :+1:

File: common/app/layout/WidthsByBreakpoint.scala (hunk 2 above)
Comment (robertberry) [inline common/app/layout/WidthsByBreakpoint.scala]: Never mind this

File: applications/app/views/fragments/galleryBody.scala.html (hunk 1 above)
Comment (rich-nguyen) [inline applications/app/views/fragments/galleryBody.scala.html]: that's a good point, I just chose a fallback to retain the imager style behaviour, but it isn't always needed. I'll leave it for this one, but the best solution would be to provide a src for the first one or two gallery items. Teensy bit more logic.

File: common/app/layout/WidthsByBreakpoint.scala (hunk 7 above)
Comment (rich-nguyen) [inline common/app/layout/WidthsByBreakpoint.scala]: We are pretty close to merging `Profile` and `BrowserWidth` together. Fist bump!

**Comment by paperboyo** [comment]
Hi,

Testing Lightbox on Win 8.1 (64-bit) Firefox 37b4, 39a1, Opera 28.0.1750.36, Chrome 42.0.2311.22 beta-m (64-bit), IE 11.0.16.

The site crashes FF 39a1 today (also reported by my colleague on 38a2 DEV edition on another machine), so that’s that...

Chrome and Opera are working splendidly, just sometimes the first image is lower res before resolving to correct one (the effect wasn’t visible before or was hidden).

FF and IE though often show lower resolution version and _never_ update it to higher res :-( It happens irrespective of how fast I browse or where in the gallery am I. Refresh resolves the issue in both browsers and loads higher resolution image.

On a separate note, https://github.com/guardian/frontend/issues/8338 still affects _all browsers_ but Chrome (cc/ @phamann, sorry if I misunderstood that this PR should fix it...)

Will be playing more...

Regards
Mateusz

**Comment by paperboyo** [comment]
Mornin'

Gallery lightbox does not work properly also on Safari 6.0.3 & Firefox 36 on 10.9.3 :-(

Exactly the same symptoms.

Regards
Mateusz

**Comment by paperboyo** [comment]
Ah, all above is true also for Article lightbox...

**Comment by paperboyo** [comment]
Hello,

Pretty please, let's not leave it like that for months :-(((

It makes lightbox barely usable: all that extra reloads just to see images exacerbate another problem (https://github.com/guardian/frontend/issues/6144) and we are back to how the old lighbox operated - it took me 8 clicks of the back button to go back from the gallery page...

Regards
Mateusz

**Comment by phamann** [comment]
@paperboyo regarding the Firefox crashing, this is a know bug with their `srcset` implementation. There is an active bugzilla report here: https://bugzilla.mozilla.org/show_bug.cgi?id=1141260

Fortunately they are aware of the issue and hopefully will not make it into a stable release for the general public.

**Comment by paperboyo** [comment]
Thanks, @phamann!

On just slightly related note. We discussed many times* the pros and cons of imagery in lightbox stretching to fit. Just another argument for stretching would be that now there is often a flash of lower resolution image before the higher res is loaded (and they aren't aligned, obviously). Would they be both stretched to fill the lightbox, it would not look so erratic, but more like progressive JPG kind of thing...

Regards
Mateusz

*(e.g. https://github.com/guardian/frontend/pull/5934#issuecomment-55766272 and my replies, arguments that @mattosborn raised there are not that much of a concern anymore, right?)

**Pull request merged**